channeldb/control_tower.go

```diff
@@ -0,0 +1,410 @@
+		if err != nil {
+			return err
+		}
+
+		// Get the existing status of this payment, if any.
+		paymentStatus := fetchPaymentStatus(bucket)
+
+		// Reset the takeoff error, to avoid carrying over an error
+		// from a previous execution of the batched db transaction.
+		takeoffErr = nil
+
+		switch paymentStatus {
+
+		// We allow retrying failed payments.
+		case StatusFailed:
+
+		// This is a new payment that is being initialized for the
+		// first time.
+		case StatusGrounded:
+
+		// We already have an InFlight payment on the network. We will
+		// disallow any new payments.
+		case StatusInFlight:
+			takeoffErr = ErrPaymentInFlight
+			return nil
+
+		// We've already completed a payment to this payment hash,
+		// forbid the switch from sending another.
+		case StatusCompleted:
+			takeoffErr = ErrAlreadyPaid
+			return nil
+
+		default:
+			takeoffErr = ErrUnknownPaymentStatus
+			return nil
+		}
+
+		// We'll move it into the InFlight state.
+		paymentStatus = StatusInFlight
+		err = bucket.Put(paymentStatusKey, StatusInFlight.Bytes())
+		if err != nil {
+			return err
+		}
+
+		// Add the payment info to the bucket, which contains the
+		// static information for this payment
+		err = bucket.Put(paymentCreationInfoKey, infoBytes)
+		if err != nil {
+			return err
+		}
+
+		// We'll delete any lingering attempt info to start with, in
+		// case we are initializing a payment that was attempted
+		// earlier, but left in a state where we could retry.
+		err = bucket.Delete(paymentAttemptInfoKey)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil
+	}
+
+	return takeoffErr
+}
+
+// RegisterAttempt atomically records the provided AttemptInfo to the DB.
+func (p *paymentControl) RegisterAttempt(paymentHash lntypes.Hash,
+	attempt *AttemptInfo) error {
+
+	// Serialize the information before opening the db transaction.
+	var a bytes.Buffer
+	if err := serializeAttemptInfo(&a, attempt); err != nil {
+		return err
+	}
+	attemptBytes := a.Bytes()
+
+	return p.db.Batch(func(tx *bbolt.Tx) error {
+		bucket, err := fetchPaymentBucket(tx, paymentHash)
+		if err != nil {
+			return err
+		}
+
+		// Add the payment attempt to the payments bucket.
+		return bucket.Put(paymentAttemptInfoKey, attemptBytes)
+	})
+}
+
+// Success transitions a payment into the Completed state. After invoking this
+// method, InitPayment should always return an error to prevent us from making
+// duplicate payments to the same payment hash. The provided preimage is
+// atomically saved to the DB for record keeping.
+func (p *paymentControl) Success(paymentHash lntypes.Hash,
+	preimage lntypes.Preimage) error {
+
+	var updateErr error
+	err := p.db.Batch(func(tx *bbolt.Tx) error {
+		bucket, err := fetchPaymentBucket(tx, paymentHash)
+		if err != nil {
+			return err
+		}
+
+		// Get the existing status, if any.
+		paymentStatus := fetchPaymentStatus(bucket)
+
+		// Reset the update error, to avoid carrying over an error
```

Comment:
Should we check here that there is a last attempt recorded? Otherwise we cannot return the route to the user and I think we always want to be able to do that.

Reply:
I'm thinking a case we finding the preimage in the preimage cache **before trying any route** is a valid outcome.

Reply:
Ok, true